bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -270,18 +192,20 @@ impl AuthenticationService {
         initial_device_name: Option<String>,
         device_id: Option<String>,
     ) -> Result<Arc<Client>, AuthenticationError> {
-        let Some(client) = self.client.read().unwrap().clone() else {
-            return Err(AuthenticationError::ClientMissing);
-        };
-
         // Login and ask the server for the full user ID as this could be different from
         // the username that was entered.
-        client.login(username, password, initial_device_name, device_id).map_err(|e| match e {
-            ClientError::Generic { msg } => AuthenticationError::Generic { message: msg },
-        })?;
-        let whoami = client.whoami()?;
-        let session =
-            client.inner.matrix_auth().session().ok_or(AuthenticationError::SessionMissing)?;
+        let client = RUNTIME.block_on(self.inner.login(
+            username,
+            password,
+            initial_device_name,
+            device_id,
+        ))?;
+        let whoami = RUNTIME
+            .block_on(client.whoami())
+            .map_err(|error| AuthenticationError::Generic { message: error.to_string() })?;
+        let session = client.matrix_auth().session().context(
+            "Login was successful but is missing a valid Session to configure the file store.",
+        )?;
 
         self.finalize_client(client, session, whoami.user_id)
```

Comment:
The main point of this refactor was to get to a stage where everything was tested so we could fix the requirement of creating a second client after login. *However!* This first step seems to make more than enough changes before trying to rationalise what to do with the 2 different types of client builder, so if it still a requirement for this part to move into the SDK, I suggest we tackle that in a follow-up PR.

As it stands right now, I've attempted to implement the AuthenticationService with the idea of using it from Rust more than the FFI use case.